crates/ra_assists/src/assists/add_explicit_type.rs

```diff
@@ -25,31 +30,85 @@ pub(crate) fn add_explicit_type(ctx: AssistCtx<impl HirDatabase>) -> Option<Assi
     let stmt = ctx.find_node_at_offset::<LetStmt>()?;
     let expr = stmt.initializer()?;
     let pat = stmt.pat()?;
+
     // Must be a binding
     let pat = match pat {
         ast::Pat::BindPat(bind_pat) => bind_pat,
         _ => return None,
     };
+
     let pat_range = pat.syntax().text_range();
     // The binding must have a name
     let name = pat.name()?;
     let name_range = name.syntax().text_range();
     // Assist not applicable if the type has already been specified
-    if stmt.syntax().children_with_tokens().any(|child| child.kind() == T![:]) {
-        return None;
+    let root = stmt.syntax();
+
+    // Found a color, we need to check if this is a placeholder def.
+    let colon: Option<SyntaxElement> =
+        root.children_with_tokens().find(|child| child.kind() == T![:]);
+
+    fn followed_by_placeholder_ty(root: &SyntaxElement) -> bool {
+        if let Some(n) = root.as_token() {
+            for sibling in n.siblings_with_tokens(Direction::Next) {
+                if let Some(sib) = sibling.as_token() {
+                    if sib.kind() == SyntaxKind::PLACEHOLDER_TYPE {
+                        return true;
+                    }
+                }
+                if let Some(sib) = sibling.as_node() {
+                    for descendent in sib.descendants_with_tokens() {
+                        if descendent.kind() == SyntaxKind::PLACEHOLDER_TYPE {
+                            return true;
+                        }
+                    }
+                }
+            }
+        }
+        false
     }
+
+    let assist_type = match colon {
+        Some(col) => {
+            if !followed_by_placeholder_ty(&col) {
+                return None;
+            }
+
+            // Colon is a token.
+            let colon_as_token = col.as_token()?;
+
+            let ty_range_start = colon_as_token.text_range().start();
+            let ty_range_end = colon_as_token
+                .siblings_with_tokens(Direction::Next)
+                .find(|s| s.kind() == SyntaxKind::EQ)
+                .expect("This is a bind expression")
+                .text_range()
+                .start();
+
+            AddExplicitTypeAssist::PlaceHolder(TextRange::from_to(ty_range_start, ty_range_end))
+        }
+        None => AddExplicitTypeAssist::Name(name_range.end()),
+    };
+
     // Infer type
     let db = ctx.db;
     let analyzer = ctx.source_analyzer(stmt.syntax(), None);
     let ty = analyzer.type_of(db, &expr)?;
+
     // Assist not applicable if the type is unknown
     if ty.contains_unknown() {
         return None;
     }
 
     ctx.add_assist(AssistId("add_explicit_type"), "add explicit type", |edit| {
         edit.target(pat_range);
-        edit.insert(name_range.end(), format!(": {}", ty.display(db)));
+
+        match assist_type {
+            AddExplicitTypeAssist::PlaceHolder(range) => {
+                edit.replace(range, format!(": {} ", ty.display(db)))
```

Comment:
Ah, so this doesn't really *fill* the placeholder, but replaces the whole type! Heh, that's neat, I haven't realised that this is poissible. 

But I think we should try to replace placeholder specifically. The original issue mentioned "walking type and pattern in lockstep", but I wonder if a more direct solution is possible. 

So, we start with `ast::BindPat`. That can be upcast via `.into` to `ast::Pat`, and that can be used with `SourceAnalyzer::type_of_pat`, which should give a type of this `_`